### DISCUSSION TITLE
Configurations in CSS? Seriously?

### DISCUSSION THREAD

**Discussion by foxeyes** [uploadcare/file-uploader#18] [Q&A]
Yes, it might look strange for the first time, but we believe it is a quite cool solution to use CSS level for configurations. Let's clarify why.

1. All uc-blocks are present in some data context. This context is provided by the element's DOM environment and could be defined by other entities. One block type can be used in different cases and data contexts (in the case of uploader, in different user workflows). This is one of their core features.
2. We have a browser-native ability for providing some abstract context down to the DOM cascade. This is a [CSS custom property](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties). They could have any custom values and are accessible inside the nested shadow roots (if you use them), unlike the other CSS properties.
3. So we have a cheap, performant, and native way to get some simple data values at any level of the DOM tree. Also, we have a convenient ability to redefine some values at any level if needed. It just works; we don't need to write some new js-library to support that process.
4. And that is pretty convenient from the developer experience perspective. You can use simple standard APIs or just a native style or class HTML attributes to define or redefine some contexts. That works with any external framework, library, or meta-platform you want to use.
5. Also, we wanted configurations to be accessible with one single request only, in the one common package with styling settings from any custom source. So you can keep or generate them by yourself at your convenience (f.e. to use the power of the Jamstack approach). By the way, that allows us to provide some configuration presets from our side, with the ability to tune something via our future UI. The same with one single resulting request.
6. If you care about your application security, it's important to have [CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) settings for your web documents. So you can't use inline scripts in your markup with more classic-style configurations anymore. That means that you need to bundle them into your code (use only one possible type and version of settings) or create another javascript file with a dedicated flow for loading sequence and caching. Not cool.

**Comment by DaniGuardiola** [comment]
This is big brained and I love it. Also, since we store our monochrome icons as SVG path data, we can replace it in a very simple way thanks to this. Great job!
